Tests/OpenCombineTests/PublisherTests/FilterTests.swift

```diff
@@ -0,0 +1,187 @@
+                                                 .value(2),
+                                                 .value(4)])
+    }
+
+    func testTryFilterCanFilterOtherFilter() {
+        // Given
+        let helper = OperatorTestHelper(publisherType: CustomPublisher.self) {
+            $0.tryFilter {
+                $0.isMultiple(of: 3)
+            }.tryFilter {
+                try nonthrowingReturn($0)
+            }
+        }
+
+        // When
+        for i in 1...9 {
+            _ = helper.publisher.send(i)
+        }
+
+        // Then
+        XCTAssertEqual(helper.tracking.history, [.subscription("TryFilter"),
+                                                 .value(3),
+                                                 .value(6),
+                                                 .value(9)])
+    }
+
+    func testTryFilterCompletesWithErrorWhenThrown() {
+        // Given
+        let helper = OperatorTestHelper(publisherType: CustomPublisher.self) {
+            $0.tryFilter {
+                try failOnFive(value: $0)
+            }
+        }
+
+        // When
+        for i in 1...5 {
+            _ = helper.publisher.send(i)
+        }
+
+        // Then
+        XCTAssertEqual(helper.tracking.history, [.subscription("TryFilter"),
+                                                 .value(1),
+                                                 .value(2),
+                                                 .value(3),
+                                                 .value(4),
+                                                 .completion(.failure(TestingError.oops))
+        ])
+    }
+
+    func testCanCompleteWithFinished() {
+        // Given
+        let helper = OperatorTestHelper(publisherType: CustomPublisher.self) {
+            $0.filter { _ in true }
+        }
+
+        // When
+        _ = helper.publisher.send(1)
+        helper.publisher.send(completion: .finished)
+
+        // Then
+        XCTAssertEqual(helper.tracking.history, [.subscription("Filter"),
+                                                 .value(1),
+                                                 .completion(.finished)])
+    }
+
+    func testCanCompleteWithError() {
+        // Given
+        let helper = OperatorTestHelper(publisherType: CustomPublisher.self) {
+            $0.filter { _ in true }
+        }
+
+        // When
+        _ = helper.publisher.send(1)
+        helper.publisher.send(completion: .failure(.oops))
+
+        // Then
+        XCTAssertEqual(helper.tracking.history, [.subscription("Filter"),
+                                                 .value(1),
+                                                 .completion(.failure(.oops))])
+    }
+
+    func testFilterSubscriptionDemand() {
+        let helper = OperatorTestHelper(publisherType: CustomPublisher.self) {
+            $0.filter { _ in true }
+        }
+
+        XCTAssertEqual(helper.subscription.history, [.requested(.unlimited)])
+    }
```

Comment:
The reason it passes like this is because `OperatorTestHelper` always requests `.unlimited`.

If we change this test to

```swift
func testFilterSubscriptionDemand() {
    let helper = OperatorTestHelper(publisherType: CustomPublisher.self,
                                    trackingDemand: .max(3),
                                    receiveValueDemand: .none) {
                                        $0.filter { $0.isMultiple(of: 2) }
                                    }

    XCTAssertEqual(helper.publisher.send(1), .max(1))
    XCTAssertEqual(helper.publisher.send(2), .max(0))
    XCTAssertEqual(helper.publisher.send(3), .max(1))
    XCTAssertEqual(helper.publisher.send(4), .max(0))
    XCTAssertEqual(helper.publisher.send(5), .max(1))
    XCTAssertEqual(helper.publisher.send(6), .max(0))
    XCTAssertEqual(helper.publisher.send(7), .max(1))
    XCTAssertEqual(helper.publisher.send(8), .max(0))

    XCTAssertEqual(helper.subscription.history, [.requested(.max(3))])
}
```

it will pass in compatibility mode, but fail with OpenCombine. The same goes for the test below.

Reply:
You can look at my attempt to handle this in `CompactMap` implementation (#32), which is practically the same as `Filter`.